modules/tensor_mechanics/src/materials/Compute2DFiniteStrain.C

```diff
@@ -18,14 +18,76 @@ validParams<Compute2DFiniteStrain>()
   InputParameters params = validParams<ComputeFiniteStrain>();
   params.addClassDescription(
       "Compute a strain increment and rotation increment for finite strains in 2D geometries.");
+
+  MooseEnum outOfPlaneDirection("x y z", "z");
+  params.addParam<MooseEnum>("out_of_plane_strain_direction",
+                             outOfPlaneDirection,
+                             "The direction of the out of of plane strain.");
   return params;
 }
 
 Compute2DFiniteStrain::Compute2DFiniteStrain(const InputParameters & parameters)
-  : ComputeFiniteStrain(parameters)
+  : ComputeFiniteStrain(parameters),
+    _out_of_plane_strain_direction(getParam<MooseEnum>("out_of_plane_strain_direction"))
 {
 }
 
+void
+Compute2DFiniteStrain::initialSetup()
+{
+  // fetch coupled variables and gradients (as stateful properties if necessary)
+  switch (_out_of_plane_strain_direction)
+  {
+    case 0: // x
+    {
+      _disp[0] = &_zero;
+      _grad_disp[0] = &_grad_zero;
+      _grad_disp_old[0] = &_grad_zero;
+      _disp[1] = &coupledValue("displacements", 0);
+      _grad_disp[1] = &coupledGradient("displacements", 0);
+      _disp[2] = &coupledValue("displacements", 1);
+      _grad_disp[2] = &coupledGradient("displacements", 1);
+      if (_fe_problem.isTransient())
+      {
+        _grad_disp_old[1] = &coupledGradientOld("displacements", 0);
+        _grad_disp_old[2] = &coupledGradientOld("displacements", 1);
+      }
+      else
+      {
+        _grad_disp_old[1] = &_grad_zero;
+        _grad_disp_old[2] = &_grad_zero;
+      }
+      break;
+    }
+    case 1: // y
+    {
+      _disp[0] = &coupledValue("displacements", 0);
+      _grad_disp[0] = &coupledGradient("displacements", 0);
+      _disp[1] = &_zero;
+      _grad_disp[1] = &_grad_zero;
+      _grad_disp_old[1] = &_grad_zero;
+      _disp[2] = &coupledValue("displacements", 1);
+      _grad_disp[2] = &coupledGradient("displacements", 1);
+      if (_fe_problem.isTransient())
+      {
+        _grad_disp_old[0] = &coupledGradientOld("displacements", 0);
+        _grad_disp_old[2] = &coupledGradientOld("displacements", 1);
+      }
+      else
+      {
+        _grad_disp_old[0] = &_grad_zero;
+        _grad_disp_old[2] = &_grad_zero;
+      }
+      break;
+    }
+    default: // z
+    {
+      ComputeIncrementalStrainBase::initialSetup();
```

Comment:
This looks like it does the right thing, but feels kind of ugly how you're calling down into the base class `initialSetup()` only for this case. Couldn't all of this code be in `ComputeIncrementalStrainBase`?

Oh yeah, now I see -- you don't want it there because the base class is not 2D specific. Hmm...

Reply:
I can move it so it always calls the base class then I overwrite for cases 0 and 1 as we discussed.